### PR TITLE
[BREAKING][FEATURE] Support decoupling contact normal forces from friction coef and sliding speed.

### LINKED ISSUE #3115: [Bug]: Incorrect tipping of object when flat surface is launched along a plane
### Bug Description

If a box is placed on a plane and an exactly horizontal velocity is applied to the box, the box should not lift off the plane, but slide along it horizontally (provided a friction coefficient mu<1 is used).

In fact, it tips slightly, causing it under a high enough velocity to lift completely from the plane and tumble.

### Steps to Reproduce

The following standalone script demonstrates the bug. The incorrect behaviour can be seen in the viewer or in an output video (attached below).
```python
import math
import numpy as np
import genesis as gs
import torch

# Time delta
DT = 0.005
# Friction coefficient
MU = 0.5
# Horizontal launch speed (m/s)
SPEED = 6.0
# Sim duration to play
SECONDS = 1.6

CAMERA_POS = (1.3, -2.2, 0.8)
CAMERA_LOOKAT = (1.3, 0.0, 0.05)
# offset kept between camera and box when tracking the moving box
CAMERA_OFFSET = tuple(p - l for p, l in zip(CAMERA_POS, CAMERA_LOOKAT))
BOX_SIZE = 0.1
BOX_START_POS = (0.0, 0.0, BOX_SIZE/2)  # start with the box sitting on the plane

# Save the render to a video file (e.g. out.mp4)
OUTPUT = None   # e.g. "out.mp4"

gs.init(backend=gs.cpu)
sc = gs.Scene(
    show_viewer=True,
    sim_options=gs.options.SimOptions(dt=DT, gravity=(0, 0, -9.81)),
    # static side camera matching the original roll_render_gen.py framing
    viewer_options=gs.options.ViewerOptions(camera_pos=CAMERA_POS,
                                            camera_lookat=CAMERA_LOOKAT,
                                            camera_fov=42),
)
# Place a box on a plane
sc.add_entity(gs.morphs.Plane())
box = sc.add_entity(gs.morphs.Box(pos=BOX_START_POS, size=(BOX_SIZE, BOX_SIZE, BOX_SIZE)),
                    surface=gs.surfaces.Rough(color=(0.85, 0.85, 0.85)))

record_cam = None
if OUTPUT is not None:
    record_cam = sc.add_camera(res=(1280, 720), pos=CAMERA_POS, lookat=CAMERA_LOOKAT, fov=42, GUI=False)

sc.build()

# Set up cameras
sc.viewer.set_camera_pose(
    pos=np.array(BOX_START_POS)+np.array(CAMERA_OFFSET),
    lookat=np.array(BOX_START_POS)
)
if record_cam is not None:
    record_cam.set_pose(
        pos=np.array(BOX_START_POS)+np.array(CAMERA_OFFSET),
        lookat=np.array(BOX_START_POS)
    )
    record_cam.start_recording()

# Friction set on the solver geoms
rs = sc.sim.rigid_solver
friction = torch.full((rs.n_geoms,), float(MU))
rs.set_geoms_friction(friction, torch.arange(rs.n_geoms))

# One settle step, then launch
sc.step()
if record_cam is not None:
    record_cam.render()

# Apply launch velocity: a purely horizontal +x launch (free-body DOFs: vx,vy,vz,wx,wy,wz).
v = box.get_dofs_velocity().clone()
v[0] = SPEED
box.set_dofs_velocity(v)

# Measure how much the box rises and tilts
max_rise = 0.0
max_tilt = 0.0
for _ in range(int(SECONDS / DT)):
    sc.step()
    q = box.get_qpos()
    q = (q if q.ndim == 1 else q[0]).detach().cpu().numpy()   # [x,y,z, qw,qx,qy,qz]
    max_rise = max(max_rise, q[2] - 0.05)
    w = max(-1.0, min(1.0, abs(float(q[3]))))
    max_tilt = max(max_tilt, math.degrees(2 * math.acos(w)))

    # keep lookat/camera z pinned to the box's starting height, follow box in x & y
    lookat = np.array([q[0], q[1], BOX_START_POS[2]])
    pos = lookat + CAMERA_OFFSET
    sc.viewer.set_camera_pose(pos=pos, lookat=lookat)
    if record_cam is not None:
        record_cam.set_pose(pos=pos, lookat=lookat)
        record_cam.render()

if record_cam is not None:
    record_cam.stop_recording(save_to_filename=OUTPUT, fps=int(round(1 / DT)))

print(f"speed={SPEED} m/s  mu={MU}  cube=0.1 m")
print(f"  peak rise = {max_rise * 1000:.1f} mm   (correct: ~0; >0 means it launched off the floor)")
print(f"  peak tilt = {max_tilt:.1f} deg  (correct: ~0; a cube should not tip at mu<1.0)")
```

### Expected Behavior

Box should slide horizontally along the plane without lifting.

### Screenshots/Videos

https://github.com/user-attachments/assets/2990e1e6-5af5-463b-8ebd-53a1b38d1559

### Relevant log output

```shell
speed=6.0 m/s  mu=0.5  cube=0.1 m
  peak rise = 318.7 mm   (correct: ~0; >0 means it launched off the floor)
  peak tilt = 179.7 deg  (correct: ~0; a cube should not tip at mu<1.0)
```

### Environment

- OS: Windows 11 26200.8457, WSL2 (6.18.33.2-microsoft-standard-WSL2)
- CPU, 13th Gen Intel(R) Core(TM) i7-1365U (1.80 GHz)
- GPU-driver version: N/A
- CUDA / CUDA-toolkit version: N/A

### Release version or Commit ID

1.2.1

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by duburcqa**
## Description

New option `contact_resolution`, selecting how a contact's normal force and friction force are resolved against each other.

`impedance` is the existing behaviour: the whole contact is one cost, and the solver may trade the normal residual against the tangential one. `signorini` instead bounds friction by the disc of radius `mu * f_n` taken from the normal force the contact has actually developed, latched at each Newton iteration. It implements the Coulomb complementarity problem eq. (C.22) of Duburcq, *Learning and Optimization of the Locomotion with an Exoskeleton for Paraplegic People*, PhD thesis, Universite Paris Sciences et Lettres, 2022 (HAL tel-04166955), Appendix C.

Defaults to `signorini` with the elliptic cone and the Newton solver, `impedance` otherwise: the pyramidal cone mixes the normal direction into every row so its rows do not separate into a normal row and a friction disc, and conjugate gradient does not reach the fixed point of the resulting successive approximation. Always `impedance` under `enable_mujoco_compatibility`.

## Related Issue

Resolves Genesis-Embodied-AI/Genesis#3115

## Motivation and Context

A 0.1 m cube given a purely horizontal 6 m/s velocity at `mu=0.5` lifts 318 mm off a flat plane and tumbles. A cube cannot tip below `mu = w / (2 * h_com) = 1`, so this is unphysical.

The friction rows' reference acceleration is pure damping, `aref_t = -b * v_t`, so at 6 m/s they demand ~1260 m/s^2 of tangential deceleration. Normal and tangential forces are chosen jointly and the cone is their only coupling, so the cheapest way to serve that demand is to buy the normal force that licenses it. Measured on a rotation-free slide, `f_n` reaches 15x-61x the weight while `f_t / f_n` stays pinned at `mu` to four digits: the solve sits on the cone boundary, and this is the minimizer rather than a convergence failure (bit-identical from 50 to 1000 iterations and from 1e-5 to 1e-9 tolerance).

The lift is the visible end of it. The same coupling makes a sliding contact's penetration depend on the friction coefficient, which it has no physical reason to: at a mere 1 m/s, penetration goes from the resting 27.3 um to -231 um at `mu=0.1` and -11388 um at `mu=0.9` (negative meaning airborne), an 11.2 mm spread across `mu` alone. Bounded against the developed normal force it stays within 0.026 um of the resting value over the same sweep.

The discriminating condition is Signorini, `0 <= delta_n PERP f_n >= 0`: on one extracted state the joint solve gives `f_n * delta_n = -5.6e+04` where the bounded one gives `-4.5e-09`.

MuJoCo shares the defect, so it cannot serve as the reference here.

## How Has This Been / Can This Be Tested?

`tests/rigid/test_friction.py::test_sliding_does_not_inflate_normal_force` is the reporter's scenario: boxes at `mu` in 0.25-0.9 launched at 2-20 m/s must hold the resting penetration to 2e-7 m, keep their normal force at the weight, and decelerate at their own `mu * g`. Since every box carries the same weight, holding all of them to one resting height pins the penetration against both the coefficient and the speed. It fails on the joint solve, where the boxes are airborne and report no contact force at all.

Beyond the suite, measured against analytical values on the reported scenario:

| | joint solve | this PR |
|---|---|---|
| rise at `mu=0.5`, 6 m/s | 311.68 mm | 0.0013 mm |
| deceleration vs analytical `mu * g` | launched | +0.000% |
| rise over 0.5-20 m/s | grows with speed | unchanged |
| sliding penetration spread over `mu` 0.1-0.9 | 11157 um | 0.026 um |
| net normal force while sliding | no contact | `1.0000 * m * g` at 2, 6 and 20 m/s |
| energy injected over the trajectory | - | exactly 0 J |
| tipping threshold | - | slides flat to `mu=0.9`, tips from `mu=1.0` |
| torsional spin-down vs analytical | -0.04% | +0.00% |

Identical results across CPU and Metal, islands on and off, torsional and rolling friction on and off, and `n_envs` 0 and 2. Step cost on identical physics (a settled pile, 118 vs 117 contacts) is **0.954x** the joint solve, the block-diagonal local Hessian being cheaper than the coupled cone it replaces.

`tests/rigid/test_friction.py` passes in full, on CPU and on Apple Metal.

Two notes on that file:

- `test_elliptic_cone_push_isotropy` is now parametrized over both resolutions. `signorini` carries a looser bound there, tracked by #3127: the box-cylinder contact manifold places its interior point at a yaw-dependent height, which gives the pusher an orientation-dependent lever arm. That is a narrow-phase defect, reproducible from `collider.detection()` alone with no solve and no dynamics, and unrelated to friction. This PR only makes it visible, by keeping the contact steadily engaged where the joint solve bounced the box and left it intermittent. Substituting a sphere for the cylinder, whose manifold is a single point, drops the spread from `4.40e-04` to `5.36e-07`, below what the joint solve achieves. The bound is tied to that issue and reverts with it.

- `test_elliptic_cone_coulomb_isotropy` gains the magnitude assertion it never had: it checked only the spread relative to its own mean, so it could not see that the mean itself was wrong. Pushed at 1.5x the Coulomb threshold with `mu=1`, the analytic sliding acceleration is `0.5 * g = 4.9050`; this PR measures `4.9052` where the joint solve measures `3.2573`, 34% low.

The second commit fixes a premature exit the first one exposes, and the cause is a metric mismatch rather than insufficient accuracy. A saturated friction block's curvature is the disc radius over the tangential residual, which collapses as the contact slides faster: measured on the reported scenario it is 620,000 to 1,800,000 times smaller than the sticking stiffness. The cost therefore has a valley orders of magnitude flatter than the stiff directions, in which the gradient norm carries no information about how far the solution still has to travel. Measured in single precision: after one iteration the gradient sits at `4.265e-06` against a tolerance of `1.450e-04`, a 34x margin, and the answer then moves by `7.2 m/s^2`.

Convergence under `signorini` therefore also requires the descent still available to be negligible, `0.5 * grad . Mgrad <= tol_scaled`. That product is twice the descent the next step would realize, and `Mgrad` already carries the inverse curvature of the current factor, which keeps it large exactly where the gradient norm looks converged, and being a property of the current iterate it reports that the step ahead is negligible rather than that the step behind was. It needs no state, no tolerance of its own since a decrement is itself a cost decrease, and no reordering: `Mgrad` is already current at the exit test on both solver arms. Step cost on the settled pile is unchanged at `0.903` of the joint solve against `0.899` before.

The other resolutions keep the single-pass test untouched: their cost holds still, and converging past the engine they are matched against is what would break consistency with it.

| BEFORE | AFTER |
| :---: | :---: |
| BEFORE | AFTER |

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Resolves SIM-294

### CODE REVIEW

File: genesis/engine/solvers/rigid/constraint/solver.py [hunk 1]
```diff
@@ -5203,7 +5303,7 @@ def func_cone_middle_cost(
     Shared by every linesearch/cost path so the coupled term is computed identically across arms.
     """
     rows_efc_D, rows_friction, con_mu, rows_jaref = _func_cone_head_load(i_c, i_b, constraint_state, rigid_config)
-    zone, N, T = _func_cone_zone(rows_jaref, con_mu, rows_friction)
+    zone, N, T = _func_cone_zone(rows_jaref, rows_efc_D, con_mu, rows_friction, rigid_config)
     c = gs.qd_float(0.0)
     if zone == 2:
         NmT = N - con_mu * T
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/solvers/rigid/constraint/solver.py]: fixed

**Comment by duburcqa** [comment]
@codex review
